crates/valence_world_time/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+
+pub struct WorldTimePlugin;
+
+impl Plugin for WorldTimePlugin {
+    fn build(&self, app: &mut bevy_app::App) {
+        app.configure_set(
+            BroadcastWorldTimeSet
+                .in_base_set(CoreSet::PostUpdate)
+                .before(WriteUpdatePacketsToInstancesSet),
+        )
+        .configure_set(
+            CalculateWorldTimeSet
+                .in_base_set(CoreSet::PostUpdate)
+                .before(BroadcastWorldTimeSet),
+        )
+        .add_systems(
+            (
+                handle_change_tracking_time_broadcast,
+                handle_interval_time_broadcast,
+            )
+                .in_set(BroadcastWorldTimeSet),
+        )
+        .add_systems(
+            (handle_linear_time_ticking, handle_linear_world_aging).in_set(CalculateWorldTimeSet),
+        )
+        .add_system(send_time_to_joined_players.before(FlushPacketsSet));
+    }
+}
+
+/// Base component for storing world time information
+#[derive(Component, Default)]
+pub struct WorldTime {
+    /// The amount of time in game tick the current world has passed
+    pub world_age: i64,
+    /// The time of day is based on the timestamp modulo 24000.
+    /// Use the [`DayPhase`] enum to easily handle common time
+    /// of day events without the need to look up information in the wiki.
+    pub time_of_day: i64,
+}
+
+impl WorldTime {
+    /// This function ensure that adding time will not resulting in
+    /// time_of_day flipping sign.
+    /// Note: If the resulting calculation set time_of_day to 0, then
+    /// the client will start advancing time.
+    pub fn add_time(&mut self, amount: i64) {
+        let client_ticking = self.client_time_ticking();
+        self.time_of_day = self.time_of_day.abs().wrapping_add(amount);
+        if self.time_of_day < 0 {
+            self.time_of_day = self.time_of_day + i64::MAX + 1;
+        }
+
+        self.set_client_time_ticking(client_ticking);
+    }
+
+    /// If the client advance world time without server update
+    pub fn client_time_ticking(&self) -> bool {
+        !self.time_of_day.is_negative()
+    }
+
+    /// Set will the client advance world time without server update
+    pub fn set_client_time_ticking(&mut self, val: bool) {
+        self.time_of_day = if val {
+            self.time_of_day.abs()
+        } else {
+            -self.time_of_day.abs()
+        };
+    }
+
+    /// Get the time part of `time_of_day`
+    pub fn current_day_time(&self) -> i64 {
+        self.time_of_day % DAY_LENGTH
+    }
+
+    /// Set the time part of `time_of_day`
+    /// Use the [`DayPhase`] enum to easily handle common time
+    /// of day events without the need to look up information in the wiki.
+    pub fn set_current_day_time(&mut self, time: i64) {
+        self.time_of_day = self.day() * DAY_LENGTH + time % DAY_LENGTH;
+    }
+
+    /// Get the current day part of `time_of_day`
+    pub fn day(&self) -> i64 {
+        self.time_of_day / DAY_LENGTH
+    }
```

Comment:
`self.time_of_day % DAY_LENGTH` and `self.time_of_day / DAY_LENGTH` return negative values when `self.time_of_day` is negative. Was that intentional?

Reply:
Some precise documentation on what exactly is returned here would be nice.

Reply:
No, that's an oversight